### ISSUE TITLE
Differences between rnn and this training script

### ISSUE THREAD

**Issue by nicholas-leonard** [pefi9/visual_attention_optim#1]
Hi @pefi9, 

So obviously the bug you were talking about could be anywhere. First thing we need to do is make sure both scripts are the same : 
- the learning rate decay might be different. I think optim uses exponential decay. rnn uses linear.
- rnn uses batch forwards, this one doesn't. https://github.com/pefi9/visual_attention_optim/blob/master/4_train.lua#L72-L83 should be removed and replaced with torch.index (which will return a batch of examples as a tensor instead of a table of tensors). Which means you can remove the for loop here https://github.com/pefi9/visual_attention_optim/blob/master/4_train.lua#L106-L124 and replace it with a single forward/backward.

I would wager that the last one will fix it as the reinforce modules and criterions do not support multiple calls to forward/backward between updates. It should be forward backward update, forward backward update and not forward backward forward backward ... update as they do in this script. I know this is how they do it in the tutorial, but it is inefficient.

**Comment by pefi9** [comment]
@nicholas-leonard thanks a lot for your suggestions. I'll give it try and let you know.
